### PR TITLE
Email Connector: Build Masking Instructions 

### LINKED ISSUE #1135: Email Connector: Generate Data for Erasure Email
### Is your feature request related to a specific problem?

https://github.com/ethyca/fidesops/issues/1134 adds the beginning pieces for an email connector.  Here we want to expand the new email connector to be able to get the data needed to populate the erasure email template.  Note the email isn't actually sent in this ticket.

### Describe the solution you'd like

- Expand EmailConnector.query_config (we can probably re-use the Manual Config)
- Stub out `EmailConnector.retrieve_data` - we're only supporting erasures at this time.
- Expand `EmailConnector.mask_data` to generate data to communicate to the user to figure out 1) *where* the relevant records are 2) *what* fields to mask and 3) instructions on *how* to mask.

_A description of what you want to happen._

### Describe alternatives you've considered, if any

_A description of any alternative solutions or features you've considered._

### Additional context

_Add any other context or screenshots about the feature request here._

### PR DESCRIPTION
**Pull request by pattisdr**
👉  The EmailConnector doesn't yet "send" an email. Documentation will be added for this feature in followup https://github.com/ethyca/fidesops/issues/1158

# Purpose

Update `EmailConnector.mask_data` to cache details about actions needed to mask that collection (how to locate the records, which fields to mask, how to mask).  

The email is not yet sent, but in #1158 we'll look into pulling these details out of the cache to send one email per "email-based dataset" at the end of the privacy request, not per collection.

# Changes
- Add `PrivacyRequest.cache_email_connector_template_contents` to cache raw details about action needed that we will email to a third party to complete.
- Add `PrivacyRequest.get_email_connector_template_contents_by_dataset` to retrieve the cached email template details.
- Pass in the original input_data into each "erasure" task that we also pass into each "access" task.  Generally, erasure tasks work with rows retrieved from the access tasks, but for email-based tasks, we won't have any data.  So we'll take the original  data we would have used to perform an access request and supply that to the third party so they can use that to locate the relevant records.
- Expand examples in contrived email dataset yaml for testing purposes
- Add a new method `TraversalNode.incoming_edges_from_same_dataset` so we know any immediate upstream dependencies of the current collection that are also in the same email-related dataset.  We won't have data from those upstream edges.
- Make some of the classes/functions used to describe manual actions needed more generic so they can be shared by email-related actions.  For example, rename PausedStep to CurrentStep, and StoppedCollection to CollectionActionRequired.  cache_restart_details was renamed to cache_action_required.

# Checklist
- [x] Update [`CHANGELOG.md`](https://github.com/ethyca/fidesops/blob/main/CHANGELOG.md) file
  - [x] Merge in main so the most recent `CHANGELOG.md` file is being appended to
  - [x] Add description within the `Unreleased` section in an appropriate category. Add a new category from the list at the top of the file if the needed one isn't already there.
  - [x] Add a link to this PR at the end of the description with the PR number as the text. example: [#1](https://github.com/ethyca/fidesops/pull/1)
- [ ] Applicable documentation updated (guides, quickstart, postman collections, tutorial, fidesdemo, [database diagram](https://github.com/ethyca/fidesops/blob/main/docs/fidesops/docs/development/update_erd_diagram.md).
- If docs updated (select one):
  - [ ] documentation complete, or draft/outline provided (tag docs-team to complete/review on this branch)
  - [ ] documentation issue created (tag docs-team to complete issue separately)
- [x] Good unit test/integration test coverage
- [ ] This PR contains a DB migration. If checked, the reviewer should confirm with the author that the [down_revision correctly references the previous migration](https://ethyca.github.io/fidesops/development/contributing_details/#alembic-migrations) before merging
- [ ] The `Run Unsafe PR Checks` label has been applied, and checks have passed, if this PR touches any external services

# Ticket

Fixes #1135

### CODE REVIEW

File: src/fidesops/ops/service/connectors/email_connector.py [hunk 1]
```diff
@@ -46,5 +48,46 @@ def mask_data(  # type: ignore
...
     ) -> Optional[int]:
-        """Stub"""
+        """Send an email to the configured user to instruct them to mask data."""
+
+        manual_action = self.build_masking_instructions(node, policy, input_data)
+        logger.info(
+            f"Manual action {manual_action.dict()} for collection {node.address.value}"
+        )
+
+        return 0
```
Comment (pattisdr) [inline src/fidesops/ops/service/connectors/email_connector.py:60]: TODO remove this, this is just for testing

File: src/fidesops/ops/service/connectors/email_connector.py [hunk 2]
```diff
@@ -46,5 +48,57 @@ def mask_data(  # type: ignore
...
+        for rule, field_paths in fields_to_mask_by_rule.items():
+            for rule_field_path in field_paths:
+                # Map field paths to the specified masking strategy.  If multiple rules target
+                # the same field path, the last one will win
+                mask_map[rule_field_path.string_path] = (
+                    rule.masking_strategy.get("strategy")
+                    if rule.masking_strategy
+                    else None
+                )
+
```
Comment (pattisdr) [inline src/fidesops/ops/service/connectors/email_connector.py:100]: I'm caching the masking strategy here, but we will probably just send a list of fields they should mask, not the strategy.  I didn't think it hurt to cache the strategy still in case that is useful later.

File: tests/ops/integration_tests/test_integration_email.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+    assert raw_email_template_values == {
+        "children": CollectionActionRequired(
+            step=CurrentStep.erasure,
+            collection=CollectionAddress("email_dataset", "children"),
+            action_needed=[
+                ManualAction(
+                    locators={
+                        "parent_id": ["email_dataset:daycare_customer:id"]
+                    },  # The only locator is on a separate collection on their end. We don't have data for it.
```
Comment (pattisdr) [inline tests/ops/integration_tests/test_integration_email.py:103]: It is tricky if one email collection is dependent on another, we won't have any data to give them to query, they need to locate on their own.  Here, we don't have any `parent_ids` to give them, they need to go to the daycare_customer collection, use the customer_id of 1, and then find the parent_id, and use that here.

File: tests/ops/integration_tests/test_integration_email.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        "payment": CollectionActionRequired(
+            step=CurrentStep.erasure,
+            collection=CollectionAddress("email_dataset", "payment"),
+            action_needed=[
+                ManualAction(
+                    locators={"payer_email": ["customer-1@example.com"]},
+                    get=None,
+                    update=None,  # Nothing to mask on this collection
+                )
+            ],
```
Comment (pattisdr) [inline tests/ops/integration_tests/test_integration_email.py:139]: Even though there's nothing to mask on the payment collection, I am still caching details for all the "email" collections because its locator may be useful in helping them find data on other downstream collections.

File: src/fidesops/ops/models/privacy_request.py [hunk 5]
```diff
@@ -71,35 +70,27 @@
 
 
 class ManualAction(BaseSchema):
-    """Surface how to manually retrieve or mask data in a database-agnostic way
+    """Surface how to retrieve or mask data in a database-agnostic way
 
     "locators" are similar to the SQL "WHERE" information.
     "get" contains a list of fields that should be retrieved from the source
-    "update" is a dictionary of fields and the value they should be replaced with.
+    "update" is a dictionary of fields and the replacement value/masking strategy
```
Comment (pattisdr) [inline src/fidesops/ops/models/privacy_request.py:77]: Reusing a structure created for the "manual" connector to cache details regarding further actions needed by some other party to continue processing of this request.

File: src/fidesops/ops/service/connectors/email_connector.py [hunk 6]
```diff
@@ -46,5 +48,56 @@ def mask_data(  # type: ignore
...
+        )
+
+        logger.info("Caching action needed for collection: '%s", node.address.value)
+        privacy_request.cache_email_connector_template_contents(
+            step=CurrentStep.erasure,
+            collection=node.address,
+            action_needed=[manual_action],
+        )
+
+        return 0  # Fidesops itself does not mask this collection.
```
Comment (pattisdr) [inline src/fidesops/ops/service/connectors/email_connector.py:68]: This is a primary change of this PR - EmailConnector.mask_data caches some instructions on how to find and mask data for this particular collection.

File: tests/ops/integration_tests/test_integration_email.py (hunk 3 above)
Comment (seanpreston) [reply tests/ops/integration_tests/test_integration_email.py:103]: Interesting edge case! @mfbrown do you have any insight on whether this is likely to happen?

File: src/fidesops/ops/service/connectors/email_connector.py (hunk 2 above)
Comment (seanpreston) [reply src/fidesops/ops/service/connectors/email_connector.py:100]: This is good — I could also imagine us precomputing values for the third party to update PII to also

**Review by seanpreston — APPROVED** [review]
